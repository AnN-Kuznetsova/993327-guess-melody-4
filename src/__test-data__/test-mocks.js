const MAX_ERRORS_COUNT = 3;

const genreQuestion = {
  type: `genre`,
  genre: `rock`,
  answers: [{
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `blues`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `jazz`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }],
};

const artistQuestion = {
  type: `artist`,
  song: {
    artist: `Third`,
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
  },
  answers: [{
    picture: `pic-one`,
    artist: `First`,
  }, {
    picture: `pic-two`,
    artist: `Second`,
  }, {
    picture: `pic-three`,
    artist: `Third`,
  }],
};


export {
  MAX_ERRORS_COUNT,
  artistQuestion,
  genreQuestion,
};
