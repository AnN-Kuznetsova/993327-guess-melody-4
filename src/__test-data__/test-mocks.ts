import {GameType, GenreQuestionType, ArtistQuestionType} from "../types";

const MAX_ERRORS_COUNT = 3;

const genreQuestion: GenreQuestionType = {
  type: GameType.GENRE,
  genre: `rock`,
  answers: [{
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `blues`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `jazz`,
  }, {
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
    genre: `rock`,
  }],
};

const artistQuestion: ArtistQuestionType = {
  type: GameType.ARTIST,
  song: {
    artist: `Third`,
    src: `https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg`,
  },
  answers: [{
    picture: `pic-one`,
    artist: `First`,
  }, {
    picture: `pic-two`,
    artist: `Second`,
  }, {
    picture: `pic-three`,
    artist: `Third`,
  }],
};


export {
  MAX_ERRORS_COUNT,
  artistQuestion,
  genreQuestion,
};
