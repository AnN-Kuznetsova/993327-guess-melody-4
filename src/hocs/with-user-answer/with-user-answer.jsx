import PropTypes from "prop-types";
import React, {PureComponent} from "react";

import {GenreQuestionsPropType} from "../../types.js";


export const withUserAnswer = (Component) => {
  class WithUserAnswer extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        answers: new Array(props.question.answers.length).fill(false),
      };

      this.handleAnswer = this.handleAnswer.bind(this);
      this.handleChange = this.handleChange.bind(this);
    }

    componentDidUpdate(prevProps) {
      const {question} = this.props;

      if (prevProps.question !== question) {
        this.setState({
          answers: new Array(question.answers.length).fill(false),
        });
      }
    }

    handleAnswer() {
      const {question, onAnswer} = this.props;
      const {answers} = this.state;

      onAnswer(question, answers);
    }

    handleChange(index, value) {
      this.setState((prevState) => {
        const userAnswers = prevState.answers.slice();
        userAnswers[index] = value;
        return {
          answers: userAnswers,
        };
      });
    }

    render() {
      const {answers} = this.state;

      return (
        <Component
          {...this.props}
          userAnswers={answers}
          onAnswer={this.handleAnswer}
          onChange={this.handleChange}
        />
      );
    }
  }


  WithUserAnswer.propTypes = {
    question: GenreQuestionsPropType.isRequired,
    onAnswer: PropTypes.func.isRequired,
  };


  return WithUserAnswer;
};
