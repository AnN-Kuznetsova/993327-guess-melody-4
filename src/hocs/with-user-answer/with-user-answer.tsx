import React from "react";
import {Subtract} from "utility-types";

import {GenreQuestionType} from "../../types";


interface Props {
  question: GenreQuestionType;
  onAnswer: (question: GenreQuestionType, answers: Answer) => void;
};

interface State {
  answers: Answer;
};

interface InjectedProps {
  userAnswers: Answer;
  onAnswer: () => void;
  onChange: (index: number) => void;
};

type Answer = boolean[];


export const withUserAnswer = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Props & Subtract<P, InjectedProps>;

  class WithUserAnswer extends React.PureComponent<T, State> {
    constructor(props) {
      super(props);

      this.state = {
        answers: new Array(props.question.answers.length).fill(false),
      };

      this.handleAnswer = this.handleAnswer.bind(this);
      this.handleChange = this.handleChange.bind(this);
    }

    componentDidUpdate(prevProps) {
      const {question} = this.props;

      if (prevProps.question !== question) {
        this.setState({
          answers: new Array(question.answers.length).fill(false),
        });
      }
    }

    handleAnswer() {
      const {question, onAnswer} = this.props;
      const {answers} = this.state;

      onAnswer(question, answers);
    }

    handleChange(index, value) {
      this.setState((prevState) => {
        const userAnswers = prevState.answers.slice();
        userAnswers[index] = value;
        return {
          answers: userAnswers,
        };
      });
    }

    render() {
      const {answers} = this.state;

      return (
        <Component
          {...this.props}
          userAnswers={answers}
          onAnswer={this.handleAnswer}
          onChange={this.handleChange}
        />
      );
    }
  }


  return WithUserAnswer;
};
