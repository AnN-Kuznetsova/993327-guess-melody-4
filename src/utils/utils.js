const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const noop = () => {
  // do nothing
};


export {
  extend,
  noop,
};
