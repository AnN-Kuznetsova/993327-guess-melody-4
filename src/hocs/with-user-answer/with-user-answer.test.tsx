import * as React from "react";
import * as renderer from "react-test-renderer";

import {withUserAnswer} from "./with-user-answer";
import {noop} from "../../utils/utils";

import {genreQuestion} from "../../__test-data__/test-mocks";


const MockComponent = (props) => {
  return (
    <div {...props} />
  );
};

const props = {
  question: genreQuestion,
  onAnswer: noop,
};

const MockComponentWithUserAnswer = withUserAnswer(MockComponent);

describe(`Render withUserAnswer`, () => {
  it(`Should match with snapshot`, () => {
    const MockComponentWithUserAnswerSnapshot = renderer.create(
        <MockComponentWithUserAnswer {...props} />
    ).toJSON();

    expect(MockComponentWithUserAnswerSnapshot).toMatchSnapshot();
  });
});
