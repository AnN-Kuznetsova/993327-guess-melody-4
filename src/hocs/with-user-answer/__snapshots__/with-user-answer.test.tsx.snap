// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render withUserAnswer Should match with snapshot 1`] = `
<div
  onAnswer={[Function]}
  onChange={[Function]}
  question={
    Object {
      "answers": Array [
        Object {
          "genre": "rock",
          "src": "https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg",
        },
        Object {
          "genre": "blues",
          "src": "https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg",
        },
        Object {
          "genre": "jazz",
          "src": "https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg",
        },
        Object {
          "genre": "rock",
          "src": "https://upload.wikimedia.org/wikipedia/commons/4/4e/BWV_543-fugue.ogg",
        },
      ],
      "genre": "rock",
      "type": "genre",
    }
  }
  userAnswers={
    Array [
      false,
      false,
      false,
      false,
    ]
  }
/>
`;
